blog/_posts/2021-07-09-FastAI-tabular-data.md

```diff
@@ -0,0 +1,207 @@
+─────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────
+   3 │    38   Private    96185   HS-grad         missing   Divorced       missing      Unmarried     Black   Fe ⋯
+                                                                                                 6 columns omitted
+
+julia> nobs(td)
+32561
+```
+
+## Data Preprocessing
+
+Although, we have loaded the data in a container which can be used later while creating a `DataLoader` and training, often we would like to perform transformations on it.  
+The tabular transformations are defined as a part of the `[DataAugmentation.jl](https://github.com/lorenzoh/DataAugmentation.jl)` package, and the currently available ones are: 
+ - Normalization
+ - FillMissing
+ - Categorify
+
+### Normalization
+
+The `Normalization` transformation involves normalizing a row of data using the mean and standard deviation of the columns. To start, we'll have to create a `Dict` which contains the required information for all rows to be normalized.
+
+```julia
+julia> using DataAugmentation, Statistics
+
+julia> continuous_cols = (:age, :fnlwgt, Symbol("education-num"), Symbol("capital-loss"), Symbol("hours-per-week"));
+
+julia> normstats = Dict();
+
+julia> for col in continuous_cols
+           normstats[col] = (Statistics.mean(skipmissing(df[:, col])), Statistics.std(skipmissing(df[:, col])))
+       end
+
+julia> normstats
+Dict{Any, Any} with 5 entries:
+  :fnlwgt                  => (1.89778e5, 105550.0)
+  :age                     => (38.5816, 13.6404)
+  Symbol("education-num")  => (10.0798, 2.573)
+  Symbol("capital-loss")   => (87.3038, 402.96)
+  Symbol("hours-per-week") => (40.4375, 12.3474)
+```
+
+After getting `normstats` dictionary, we can create the `NormalizeRow` object which will help us perform this transformation.
+
+
+```julia
+julia> normalize = DataAugmentation.NormalizeRow(normstats, continuous_cols);
+```
+
+Now let's quickly get a row of data and see `NormalizeRow` in action.
+
+All the transformations work on `TabularItem` objects and use the `apply` function for transforming the data.
+
+```julia
+julia> row = getobs(td, 1)
+DataFrameRow
+ Row │ age    workclass  fnlwgt  education    education-num  marital-status       occupation  relationship  race ⋯
+     │ Int64  String     Int64   String       Float64?       String               String?     String        Stri ⋯
+─────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────
+   1 │    49   Private   101320   Assoc-acdm           12.0   Married-civ-spouse  missing      Wife          Whi ⋯
+                                                                                                 7 columns omitted
+
+julia> item = DataAugmentation.TabularItem(row, Tables.columnnames(df));
+
+julia> DataAugmentation.apply(normalize, item).data
```

Comment:
Show the `TablularItem` here to clarify what is written in the next sentence? We never see the `TabularItem` post normalisation.